src/common/ceph_clock.h

```diff
@@ -0,0 +1,420 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * Author: Shinobu Kinjo <shinobu@redhat.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+/**
+ * @Param:
+ *  *_start_  construction point
+ *   ** This is to avoid any noise
+ *  *_now_    start point
+ *   ** This is the point to calculate the latency
+ *  *_end_    end point
+ *
+ * @Usage:
+ *  // Get timestamp 
+ *  ceph_clock cc;
+ *  auto t1 = cc.now<utime_t>()
+ *  auto t2 = cc.now<mono_time>();
+ *  auto t3 = cc.now<corse_mono_time>();
+ *
+ *  // Get duration
+ *  ceph::mono_clock::duration second;
+ *  cc.set_duration<seconds, coarse_mono_time>(second, 1_s);
+ *
+ *  // Get count
+ *  auto c1 = cc.get_count<seconds, coarse_mono_time>();
+ */
+#ifndef CEPH_CLOCK_H
+#define CEPH_CLOCK_H
+
+#include "common/ceph_time.h"
+#include "include/utime.h"
+#include <type_traits>
+#include <time.h>
+#include <boost/noncopyable.hpp>
+
+using namespace std::chrono;
+
+constexpr unsigned long long operator"" _ns (unsigned long long ns) {
+  return ns / 1000000000;
+}
+constexpr unsigned long long operator"" _ms (unsigned long long ms) {
+  return ms / 1000;
+}
+constexpr unsigned long long operator"" _s (unsigned long long s) {
+  return s;
+}
+constexpr unsigned long long operator"" _m (unsigned long long m) {
+  return m * 60;
+}
+
+namespace ceph {
+
+struct ceph_clock : private boost::noncopyable {
+  ceph_clock() :
+    mt_start_(ceph::mono_clock::now()),
+    cmt_start_(ceph::coarse_mono_clock::now()),
+    ut_start_(ceph_clock_now_()) {} 
```

Comment:
as i mentioned in the previous pr, we really don't want to read from all 3 clocks in this constructor. this means that the `ceph_clock_now()` function that uses this (`ceph_clock cc; return cc.now<utime_t>();`) will read from 4 clocks. surely this doesn't accomplish the goal:

> 3. Minimize migration cost as much as possible